pandas/core/strings.py

```diff
@@ -505,6 +507,280 @@ def f(x):
+    be used for column names; otherwise capture group numbers will be
+    used. The dtype of each result column is always object, even when
+    no match is found.
+
+    Examples
+    --------
+    A pattern with one group will return a DataFrame with one column.
+    Non-matches will be NaN.
+
+    >>> s = Series(['a1', 'b2', 'c3'])
+    >>> s.str.extract('[ab](\d)')
+         0
+    0    1
+    1    2
+    2  NaN
+
+    A pattern with two groups will return a DataFrame with two columns.
+
+    >>> s.str.extract('([ab])(\d)')
+         0    1
+    0    a    1
+    1    b    2
+    2  NaN  NaN
+
+    A pattern may contain optional groups.
+
+    >>> s.str.extract('([ab])?(\d)')
+         0  1
+    0    a  1
+    1    b  2
+    2  NaN  3
+
+    Named groups will become column names in the result.
+
+    >>> s.str.extract('(?P<letter>[ab])(?P<digit>\d)')
+      letter digit
+    0      a     1
+    1      b     2
+    2    NaN   NaN
+
+    """
+    from pandas.core.series import Series
+    from pandas.core.frame import DataFrame
+    from pandas.core.index import Index
+
+    regex = re.compile(pat, flags=flags)
+
+    if regex.groups == 0:
+        raise ValueError("pattern contains no capture groups")
+
+    names = dict(zip(regex.groupindex.values(), regex.groupindex.keys()))
+    columns = [names.get(1 + i, i) for i in range(regex.groups)]
+    if len(arr) == 0:
+        return DataFrame(columns=columns, dtype=object)
+
+    empty_row = [np.nan] * regex.groups
+
+    def f(x):
+        if not isinstance(x, compat.string_types):
+            return empty_row
+        m = regex.search(x)
+        if m:
+            return [np.nan if item is None else item for item in m.groups()]
+        else:
+            return empty_row
+    try:
+        result_index = arr.index
+    except AttributeError:
+        result_index = None
+    return DataFrame(
+        [f(val) for val in arr],
+        columns=columns,
+        index=result_index,
+        dtype=object)
+
+
+def str_extract(arr, pat, flags=0, expand=False):
+    """
+    For each subject string in the Series, extract groups from the
+    first match of regular expression pat.
+
+    .. versionadded:: 0.13.0
+
+    Parameters
+    ----------
+    pat : string
+        Regular expression pattern with capturing groups
+    flags : int, default 0 (no flags)
+        re module flags, e.g. re.IGNORECASE
+    expand : bool, default False
+        * If True, return DataFrame.
+        * If False, return Series/Index/DataFrame.
+
+    Returns
+    -------
+    DataFrame with one row for each subject string, and one column for
+    each group. Any capture group names in regular expression pat will
+    be used for column names; otherwise capture group numbers will be
+    used. The dtype of each result column is always object, even when
+    no match is found. If expand=True and pat has only one capture group,
+    then return a Series (if subject is a Series) or Index (if subject
+    is an Index).
+
```

Comment:
Maybe also add a 'See also' in this docstring that refers to extractall (and vice versa there)